pkg/kubecost/totals.go

```diff
@@ -0,0 +1,367 @@
+	if shareSplit == ShareNone {
+		return 0.0, 0.0, 0.0
+	}
+
+	if shareSplit != ShareEven {
+		shareSplit = ShareWeighted
+	}
+
+	var cpuCoeff, gpuCoeff, ramCoeff float64
+
+	if _, ok := allocationTotals[key]; !ok {
+		return 0.0, 0.0, 0.0
+	}
+
+	if shareSplit == ShareEven {
+		coeff := 1.0 / float64(allocationTotals[key].Count)
+		return coeff, coeff, coeff
+	}
+
+	if allocationTotals[key].CPUCost > 0 {
+		cpuCoeff = cpuCost / allocationTotals[key].CPUCost
+	}
+
+	if allocationTotals[key].GPUCost > 0 {
+		gpuCoeff = cpuCost / allocationTotals[key].GPUCost
+	}
+
+	if allocationTotals[key].RAMCost > 0 {
+		ramCoeff = ramCost / allocationTotals[key].RAMCost
+	}
+
+	return cpuCoeff, gpuCoeff, ramCoeff
+}
+
+type ResourceTotalsStore interface {
+	GetResourceTotalsByCluster(start, end time.Time) map[string]*ResourceTotals
+	GetResourceTotalsByNode(start, end time.Time) map[string]*ResourceTotals
+	SetResourceTotalsByCluster(start, end time.Time, rts map[string]*ResourceTotals)
+	SetResourceTotalsByNode(start, end time.Time, rts map[string]*ResourceTotals)
+}
+
+func UpdateResourceTotalsStoreFromAllocationSet(rts ResourceTotalsStore, as *AllocationSet) error {
+	if rts == nil {
+		return errors.New("cannot update resource totals from AllocationSet for nil ResourceTotalsStore")
+	}
+
+	if as == nil {
+		return errors.New("cannot update resource totals from AllocationSet for nil AllocationSet")
+	}
+
+	start := *as.Window.Start()
+	end := *as.Window.End()
+
+	rtsByCluster := ComputeResourceTotalsFromAllocations(as, AllocationClusterProp)
+	rts.SetResourceTotalsByCluster(start, end, rtsByCluster)
+
+	rtsByNode := ComputeResourceTotalsFromAllocations(as, AllocationNodeProp)
+	rts.SetResourceTotalsByNode(start, end, rtsByNode)
+
+	log.Infof("ETL: Allocation: updated resource totals: %s", as.Window)
+
+	return nil
+}
+
+func UpdateResourceTotalsStoreFromAssetSet(rts ResourceTotalsStore, as *AssetSet) error {
+	if rts == nil {
+		return errors.New("cannot update resource totals from AssetSet for nil ResourceTotalsStore")
+	}
+
+	if as == nil {
+		return errors.New("cannot update resource totals from AssetSet for nil AssetSet")
+	}
+
+	start := *as.Window.Start()
+	end := *as.Window.End()
+
+	rtsByCluster := ComputeResourceTotalsFromAssets(as, AssetClusterProp)
+	rts.SetResourceTotalsByCluster(start, end, rtsByCluster)
+
+	rtsByNode := ComputeResourceTotalsFromAssets(as, AssetNodeProp)
+	rts.SetResourceTotalsByNode(start, end, rtsByNode)
+
+	log.Infof("ETL: Asset: updated resource totals: %s", as.Window)
+
+	return nil
+}
+
+type MemoryResourceTotalsStore struct {
+	byCluster *cache.Cache
+	byNode    *cache.Cache
+}
+
+func NewResourceTotalsStore() *MemoryResourceTotalsStore {
+	return &MemoryResourceTotalsStore{
+		byCluster: cache.New(cache.NoExpiration, cache.NoExpiration),
+		byNode:    cache.New(cache.NoExpiration, cache.NoExpiration),
```

Comment:
Should these really have no expiration? Perhaps it should be set to the maximum ETL coverage?

Reply:
What's the pathological case where you're concerned about no expiration? It's not clear to me how long we'd have to wait before it was better to have no data than the last-updated data.

Reply:
Say we've got a user with a 90d ETL and the pod manages to stay alive for 180+ days. Does this cache grow unbounded, eventually causing an OOM crash?

Reply:
Ah, very fair point. This does save me from having to manually "roll off the back". I'm convinced.